### PR TITLE
19834 replace prepare for user handler with graph stage agolubev

### PR DESCRIPTION
**Pull request by agolubev**
Ref #19834

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala [hunk 1]
```diff
@@ -64,20 +65,24 @@ private[http] object WebSocket {
...
-          else if (inMessage) ctx.fail(new ProtocolException(s"Truncated message, peer closed connection in the middle of message."))
-          else ctx.finish()
-        case ActivelyCloseWithCode(code, reason) ⇒
-          if (code.exists(Protocol.CloseCodes.isError)) ctx.fail(new ProtocolException(s"Closing connection with error code $code"))
-          else ctx.fail(new IllegalStateException("Regular close from FrameHandler is unexpected"))
-        case x: MessageDataPart ⇒
-          inMessage = !x.last
-          ctx.push(x)
-        case x ⇒ ctx.push(x)
+    class PrepareForUserHandler extends SimpleLinearGraphStage[MessagePart] {
```
Comment (viktorklang) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala]: Does this need to be a class, couldn't it be defined as an `object` outside of this method?

File: akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala [hunk 2]
```diff
@@ -113,7 +118,7 @@ private[http] object WebSocket {
 
     def prepareMessages: Flow[MessagePart, Message, NotUsed] =
       Flow[MessagePart]
-        .transform(() ⇒ new PrepareForUserHandler)
+        .via(new PrepareForUserHandler)
         .splitWhen(_.isMessageEnd) // FIXME using splitAfter from #16885 would simplify protocol a lot
```
Comment (viktorklang) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala]: this FIXME should be addressible: https://github.com/akka/akka/issues/16885

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala]: I'll make it more rememberable + discoverable during planning as a ticket.

File: akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala (hunk 1 above)
Comment (agolubev) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala]: it has no state so I will convert it to an object

File: akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala (hunk 2 above)
Comment (agolubev) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala]: ticket  #20149 created

**Comment by agolubev** [comment]
changed according to feedback. ready for review.

**Comment by viktorklang** [comment]
LGTM, thanks @agolubev!

**Comment by agolubev** [comment]
Thanks @viktorklang!
Now looking for another LGTM.

File: akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala [hunk 3]
```diff
@@ -7,6 +7,7 @@ package akka.http.impl.engine.ws
 import java.util.Random
 import akka.NotUsed
 import akka.event.LoggingAdapter
+import akka.stream.impl.fusing.GraphStages.SimpleLinearGraphStage
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala]: I think I'd personally rather have a bit of repeated code using the public API than introduce a dependency on stream internal here.

**Comment by johanandren** [comment]
LGTM except for the dependency on akka stream internals

**Comment by agolubev** [comment]
@johanandren, let me change dependency to GraphStage. I'll rebase and squash then

**Comment by patriknw** [comment]
LGTM

**Pull request merged**